test/test_datapipe.py

```diff
@@ -1554,68 +1547,56 @@ def test_stream_reader_iterdatapipe(self):
-                    multiprocessing_context=mp_ctx,
-                    worker_init_fn=_worker_init_fn,
-                    persistent_workers=pw
-                )
-
-                # No seed
-                dl_res_ns = list(dl)
-                self.assertEqual(len(dl_res_ns), len(exp))
-                self.assertEqual(sorted(dl_res_ns), sorted(exp))
-
-                # Same seeds
-                dl_res = []
-                for epoch in range(2):
-                    torch.manual_seed(123)
-                    dl_res.append(list(dl))
-                self.assertEqual(dl_res[0], dl_res[1])
-
-                # Different seeds
-                torch.manual_seed(321)
-                dl_res.append(list(dl))
-
-                self.assertEqual(len(dl_res[0]), len(dl_res[2]))
-                self.assertNotEqual(dl_res[0], dl_res[2])
-                self.assertEqual(sorted(dl_res[0]), sorted(dl_res[2]))
-
-
-        shuffle_dp_nl = IDP_NoLen(range(20)).shuffle(buffer_size=5)
-        with self.assertRaisesRegex(TypeError, r"instance doesn't have valid length$"):
-            len(shuffle_dp_nl)
+            shuffle_dp = input_dp.shuffle(buffer_size=0)
+
+        # Functional Test: No seed
+        shuffler_dp = input_dp.shuffle()
+        self.assertEqual(set(range(10)), set(shuffler_dp))
+
+        # Functional Test: With global seed
+        torch.manual_seed(123)
+        shuffler_dp = input_dp.shuffle()
+        res = list(shuffler_dp)
+        torch.manual_seed(123)
+        self.assertEqual(list(shuffler_dp), res)
+
+        # Functional Test: Set seed
+        shuffler_dp = input_dp.shuffle().set_seed(123)
+        res = list(shuffler_dp)
+        shuffler_dp.set_seed(123)
+        self.assertEqual(list(shuffler_dp), res)
+
+        # Functional Test: deactivate shuffling via set_shuffle
+        unshuffled_dp = input_dp.shuffle().set_shuffle(False)
+        self.assertEqual(list(unshuffled_dp), list(input_dp))
+
+        # Reset Test:
+        shuffler_dp = input_dp.shuffle()
+        n_elements_before_reset = 5
+        res_before_reset, res_after_reset = reset_after_n_next_calls(shuffler_dp, n_elements_before_reset)
+        self.assertEqual(5, len(res_before_reset))
+        for x in res_before_reset:
+            self.assertTrue(x in set(range(10)))
+        self.assertEqual(set(range(10)), set(res_after_reset))
 
-        # Test: deactivate shuffling via set_shuffle
-        unshuffled_dp = input_ds.shuffle().set_shuffle(False)
-        self.assertEqual(list(unshuffled_dp), list(input_ds))
+        # __len__ Test: returns the length of the input DataPipe
+        shuffler_dp = input_dp.shuffle()
+        self.assertEqual(10, len(shuffler_dp))
+        exp = list(range(100))
+
+        # Serialization Test
+        shuffler_dp = input_dp.shuffle()
+        exp = []
+        it = iter(shuffler_dp)
+        for _ in range(2):
+            exp.append(next(it))
+        shuffler_dp_copy = pickle.loads(pickle.dumps(shuffler_dp))
```

Comment:
Without state transition within `__setstate__`, here `shuffler_dp` has snapshot state as `Iterating` but `shuffler_dp_copy` has the snapshot state as `NotStarted`. Then `shuffler_dp_copy` would do another `reset`, which makes serialization/deserialization useless.

Reply:
Is this test case what we want? You serialize/deserialize a DataPipe that has been partially read, but the expected result is that the copied DataPipe will return the full result (starting from the beginning).

When does this occur? Are DataPipes partially read prior to the start of multiprocessing?

Two somewhat related observations:
1) If `set_seed` is called prior to pickling, then the copied DP will have the seed which can be used in the next iteration as .
2) If we save the `buffer` and `rng_state (as we plan to later on), then the copied will also only have the rest of the elements (but not the first two).

Reply:
I am sorry that my test case is misleading. It should be something like:
```py
shuffler_dp = input_dp.shuffle()
it = iter(shuffler_dp)
for _ in range(2):
        _ = next(it)
shuffler_dp_copy = pickle.loads(pickle.dumps(shuffler_dp))
exp = list(it)
res = list(shuffler_dp_copy)
self.assertEqual(exp, res)
```
Does this make more sense? Then, I need to serialize both `buffer` and `rng_state` to achieve so.

Reply:
Yea, I think the only way this test case can pass is, as you describe, serialize/deserialize `buffer` and `rng_state`.

The alternate design would be to call some snapshot restoration functions (e.g. simple snapshot restoration).

Reply:
Revise the test case